src/helpers/buffers.rs

```diff
@@ -131,3 +194,231 @@ impl ReceiveBuffer {
+            offset: initial_offset,
+            batch_size,
+            accepted_range: RecordId::from(initial_offset * batch_size)
+                ..RecordId::from(initial_offset * batch_size + batch_size),
+            len: 0,
+            data: {
+                let batch_size = batch_size as usize;
+                let mut data = Vec::with_capacity(batch_size);
+                data.resize_with(batch_size, || None);
+
+                data
+            },
+        }
+    }
+
+    pub fn push(&mut self, msg: MessageEnvelope) -> Result<(), SendBufferError> {
+        let idx = self.accept_msg(&msg)? as usize;
+        if self.data[idx].is_some() {
+            Err(SendBufferError::DuplicateMessage {
+                record_id: msg.record_id,
+            })
+        } else {
+            self.data[idx] = Some(msg);
+            self.len += 1;
+
+            Ok(())
+        }
+    }
+
+    pub fn is_full(&self) -> bool {
+        self.len == self.batch_size
+    }
+
+    pub fn flush(&mut self) -> Vec<MessageEnvelope> {
+        let offset = self.offset + 1;
+        let batch_size = self.batch_size;
+        let prev_self = mem::replace(self, Self::new(offset, batch_size));
+
+        prev_self.into()
+    }
+
+    fn accept_msg(&self, msg: &MessageEnvelope) -> Result<u32, SendBufferError> {
+        if self.accepted_range.contains(&msg.record_id) {
+            // safety: safe to subtract unsigned as record_id is greater than min
+            Ok(u32::from(msg.record_id) - u32::from(self.accepted_range.start))
+        } else {
+            Err(SendBufferError::RecordOutOfRange {
+                record_id: msg.record_id,
+                accepted_range: self.accepted_range.clone(),
+            })
+        }
+    }
+}
+
+impl From<SendBatch> for Vec<MessageEnvelope> {
+    fn from(value: SendBatch) -> Self {
+        // we can safely reinterpret because batch is full
+        if value.len == value.batch_size {
+            // what we are about to do is unsafe, so it is better to check that we won't cause a UB
+            // Note that if the implementation is correct, checking self.len == self.batch_size is enough
+            // to prevent undefined behaviour, this is just an extra safety guard
+            debug_assert!(value.data.as_slice().iter().all(Option::is_some));
+
+            // safety: Option<T> and T have the same layout as long as Option is Some
+            // https://doc.rust-lang.org/std/option/#representation
+            // we checked above that all elements inside the vec are set to Some(T) by verifying the len
+            unsafe {
+                // don't drop the original vector
+                let mut data = ManuallyDrop::new(value.data);
+
+                // reinterpret Vec<Option<T>> as Vec<T>
+                Vec::from_raw_parts(
+                    data.as_mut_ptr().cast::<MessageEnvelope>(),
+                    data.len(),
+                    data.capacity(),
+                )
+            }
```

Comment:
Could this potentially change in some future version of Rust? I mean, the memory layout? Doesn't this approach kind of bind us to an implementation detail?

Reply:
I don't think so because it is documented behaviour and in many places in standard library it is expected for `Option<T>` to have the same layout as `T`. Changing this would likely be a major breaking change so I don't think they will even consider that

Reply:
I don't think that that is strictly true.  `Option<T>` can require a different layout than `T`, though for some types it can share a layout (and that is the general preference where possible).  For instance, `Option<u32>` cannot take just 4 bytes as it has $2^32+1$ different possible values, though `Option<&T>` as a reference type can always use a null value to indicate `None`.